src/server/frontend_wayland/wl_surface.cpp

```diff
@@ -289,14 +292,53 @@ void mf::WlSurface::set_input_region(std::experimental::optional<wl_resource*> c
 {
     if (region)
     {
-        // since pending.input_shape is an optional optional, this is needed
         auto shape = WlRegion::from(region.value())->rectangle_vector();
-        pending.input_shape = decltype(pending.input_shape)::value_type{move(shape)};
+
+        // only set the shape if it is different from the currently committed input shape
+        bool shape_changed;
+        if (!input_shape || input_shape.value().size() != shape.size())
+        {
+            shape_changed = true;
+        }
+        else
+        {
+            // The same input shape could be represented by the same rectangles in a different order, or event
+            // different rectangles. We don't check for that, however, because it would only cause an unnecessary
+            // update and not do any real harm. Checking for identical vectors should cover most cases.
+            shape_changed = false;
+            for (unsigned i = 0; i < shape.size(); i++)
+            {
+                if (input_shape.value()[i] != shape[i])
+                {
+                    shape_changed = true;
+                    break;
```

Comment:
Is the complexity of short-circuiting this loop really so much better than the simpler `shape_changed = input_shape.value() != shape`?

Reply:
For some reason I had it in my mind that `std::vector` didn't implement `==`. Fixed.